tools/DynamoPackagesAnalyzer/PackageSource/ZipArchiveSource.cs

```diff
@@ -0,0 +1,132 @@
+using System.Collections.Concurrent;
+using DynamoPackagesAnalyzer.Helper;
+using DynamoPackagesAnalyzer.Models;
+using DynamoPackagesAnalyzer.Models.CommandLine;
+using DynamoPackagesAnalyzer.Models.Greg;
+using Microsoft.Extensions.Configuration;
+using static DynamoPackagesAnalyzer.Helper.LogHelper;
+
+namespace DynamoPackagesAnalyzer.PackageSource
+{
+    /// <summary>
+    /// Provides method to analize package zip archives 
+    /// </summary>
+    internal class ZipArchiveSource
+    {
+        private readonly ZipArchiveOptions options;
+        private readonly FileInfo[] files;
+        private readonly DirectoryInfo workspace;
+        private readonly IConfigurationRoot configuration;
+        private PackageHeader packageHeader;
+
+        public ZipArchiveSource(ZipArchiveOptions options, PackageHeader packageHeader = null, DirectoryInfo workspace = null)
+        {
+            this.options = options;
+            files = options.Files.Select(f => new FileInfo(f)).ToArray();
+
+            if (!files.All(f => f.Exists))
+            {
+                throw new InvalidOperationException($"{string.Join(", ", files.Where(f => !f.Exists))} does not exists");
+            }
+
+            if (!files.All(f => f.Extension.Equals(".zip")))
+            {
+                throw new InvalidOperationException($"Some files does not have a valid zip extension: {string.Join(", ", files.Where(f => !f.Extension.Equals(".zip")))}");
+            }
+            this.packageHeader = packageHeader;
+            this.workspace = workspace ?? WorkspaceHelper.GetWorkspace();
+            configuration = ConfigHelper.GetConfiguration();
+        }
+
+        /// <summary>
+        /// Starts the analysis process on the zip files provided
+        /// </summary>
+        /// <returns></returns>
+        public async Task<List<AnalyzedPackage>> RunAnalysis()
+        {
+            ConcurrentBag<AnalyzedPackage> packages = new ConcurrentBag<AnalyzedPackage>();
+
+            switch (options.MultiThread)
+            {
+                case true:
+                    int MaxDegreeOfParallelism = int.Parse(configuration["MaxDegreeOfParallelism"]);
+                    await Parallel.ForEachAsync(files,
+                        new ParallelOptions { MaxDegreeOfParallelism = MaxDegreeOfParallelism },
+                        async (item, cancellationToken) =>
+                        {
+                            packages.AddRange(await Process(item));
+                        }
+                    );
+                    break;
+                default:
+                    foreach (FileInfo item in files)
+                    {
+                        packages.AddRange(await Process(item));
+                    }
+                    break;
+            }
+
```

Comment:
Any reason why you prefer a switch-case statement over a simple if-then for true-false conditions?

Reply:
As a personal preference, when the if statement may have several lines of code or I have to return values, a switch statement would be easier to read

Reply:
Not sure about that (easier to read), which is why I asked the question. ;)